src/keyDerivation.c

```diff
@@ -88,3 +94,178 @@ void extractRawPublicKey(
+{
+	// other checks are when deriving private key
+	VALIDATE_PARAM(pathLength >= 5);
+	VALIDATE_PARAM(bip32Path[3] == 0 || bip32Path[3] == 1);
+}
+
+
+// pub_key + chain_code
+void deriveExtendedPublicKey(
+        const uint32_t* bip32Path, uint32_t pathLength,
+        uint8_t* out, size_t outputSize
+)
+{
+	cx_ecfp_public_key_t publicKey;
+	privateKey_t privateKey;
+	chain_code_t chainCode;
+
+	BEGIN_TRY {
+		TRY {
+			derivePrivateKey(
+			        bip32Path,
+			        pathLength,
+			        &chainCode,
+			        &privateKey
+			);
+
+			deriveRawPublicKey(&privateKey, &publicKey);
+		}
+		FINALLY {
+			os_memset(&privateKey, 0, SIZEOF(privateKey));
+		}
+	} END_TRY;
+
+
+	uint8_t rawPublicKey[PUBLIC_KEY_SIZE];
+	extractRawPublicKey(&publicKey, rawPublicKey, SIZEOF(rawPublicKey));
+
+
+	ASSERT(outputSize = CHAIN_CODE_SIZE + PUBLIC_KEY_SIZE);
+	os_memmove(out, rawPublicKey, CHAIN_CODE_SIZE);
+	os_memmove(out + CHAIN_CODE_SIZE, chainCode.code, PUBLIC_KEY_SIZE);
+}
+
+void addressRootFromExtPubKey(
+        const uint8_t* extPubKey, size_t extPubKeySize,
+        uint8_t* addressRoot, size_t addressRootSize
+)
+{
+	ASSERT(extPubKeySize == EXTENDED_PUBKEY_SIZE);
+	ASSERT(addressRootSize == 28); //
+
+	uint8_t cborBuf[64 + 10];
+	uint8_t* ptr = cborBuf;
+	uint8_t* end = END(cborBuf);
+
+#define WRITE_TOKEN(type, value) \
+	{ \
+	    ptr += cbor_writeToken(type, value, ptr, end - ptr); \
+	}
+#define WRITE_DATA(buf, bufSize) \
+	{ \
+	    if (ptr + bufSize > end) THROW(ERR_DATA_TOO_LARGE); \
+	    os_memmove(ptr, buf, bufSize); \
+	    ptr += bufSize; \
+	}
+
+	// [0, [0, publicKey:chainCode], Map(0)]
+	// TODO(ppershing): what are the first two 0 constants?
+	WRITE_TOKEN(CBOR_TYPE_ARRAY, 3);
+	WRITE_TOKEN(CBOR_TYPE_UNSIGNED, 0);
+
+	// enter inner array
+	WRITE_TOKEN(CBOR_TYPE_ARRAY, 2);
+	WRITE_TOKEN(CBOR_TYPE_UNSIGNED, 0);
+	WRITE_TOKEN(CBOR_TYPE_BYTES, extPubKeySize);
+	WRITE_DATA(extPubKey, extPubKeySize);
+	// exit inner array
+
+	WRITE_TOKEN(CBOR_TYPE_MAP, 0);
+
+#undef WRITE_TOKEN
+#undef WRITE_DATA
+
+	// cborBuf is hashed twice. First by sha3_256 and then by blake2b_224
+	uint8_t cborShaHash[32];
+	sha3_256_hash(
+	        cborBuf, ptr - cborBuf,
+	        cborShaHash, SIZEOF(cborShaHash)
+	);
+	blake2b_224_hash(
+	        cborShaHash, SIZEOF(cborShaHash),
+	        addressRoot, addressRootSize
+	);
+}
+
+uint32_t deriveAddress(
+        const uint32_t* bip32Path, uint32_t pathLength,
+        uint8_t* address, size_t maxSize
+)
+{
+	validatePathForAddressDerivation(bip32Path, pathLength);
+
+	uint8_t addressRoot[28];
+	{
+		uint8_t extPubKey[EXTENDED_PUBKEY_SIZE];
+
+		deriveExtendedPublicKey(
+		        bip32Path, pathLength,
+		        extPubKey, SIZEOF(extPubKey)
+		);
+
+		addressRootFromExtPubKey(
+		        extPubKey, SIZEOF(extPubKey),
+		        addressRoot, SIZEOF(addressRoot)
+		);
+	}
+
+	uint8_t addressRaw[100];
+	uint8_t* ptr = BEGIN(addressRaw);
+	uint8_t* end = END(addressRaw);
+
+#define WRITE_TOKEN(type, value) \
```

Comment:
this is duplicite definition of WRITE_TOKEN

Reply:
Yep, it is duplicate. But given the fact that ptr and end are different variables in the two cases (because it is different function(, I expected it to be safer to redefine it.
Maybe we should take them as parameters though